File: 2017-ag-precip/ag_buildSamples.m
% load cropWxData (produced by ag_mergeCropWxData.m) and select the parts
% of the wx data that match the samples here

cropDataBaseDir = 'f:/data/ag/crop/';

tempPrc = 99.9;
precipPrc = 99.9;

pcnt = 0;
tcnt = 0;

% whether to generate full dataset with group and original hourly wx data
% or just groups and seasonal summaries
full = false;

% find names of all state crop data files
txtFileNames = dir([cropDataBaseDir, '/*.mat']);
txtFileNames = {txtFileNames.name};

% loop over states
for s = 1:length(txtFileNames)
    
    curFileParts = strsplit(txtFileNames{s}, '.');
    
    ['loading ' txtFileNames{s} '...']
    cropWxData = load([cropDataBaseDir txtFileNames{s}]);
    cropWxData = cropWxData.cropWxData;
    
    % loop over counties
    for c = 1:length(cropWxData{3})
        
        % if county doesn't have matched wx data, skip it
        if length(cropWxData{3}{c}) == 5
            cropWxData{3}{c}{5} = {};
            continue;
        end
        
        ['processing ' cropWxData{1} '/' cropWxData{3}{c}{1} '...']
        
        temp = cropWxData{3}{c}{7}{1}{end-2};
        precip = cropWxData{3}{c}{7}{1}{end};

        years = cropWxData{3}{c}{7}{1}{4};
        months = cropWxData{3}{c}{7}{1}{5};

        % select only growing season from april - sept
        growingSeasonInd = find(months>= 4 & months <= 9);

        if length(growingSeasonInd) == 0
            continue;
        end
        
        temp = temp(growingSeasonInd);
        precip = precip(growingSeasonInd);
        years = years(growingSeasonInd);
        months = months(growingSeasonInd);

        % calculate seasonal means
        tempMean = [];
        precipSum = [];
        
        tempThresh = prctile(temp, tempPrc);
        precipThresh = prctile(precip, tempPrc);

        tempGroup = [];
        precipGroup = [];
        
        % the indicies that mark the beginning of each year in the wx data
        yearInd = [];
        
        % find groups
        lastYear = years(1);
        lastYearStartInd = 1;
        lastYearEndInd = 1;
        
        % first dim = index
        yearInd(1, 1) = 1;
        % second dim = year
        yearInd(1, 2) = years(1);
        
        % loop over all years
        for y = 1:length(years)
            
            % if we are on a new year, use collected indicies to select
            % temp/precip data for this year
            if years(y) ~= lastYear && y <= length(temp) && y <= length(precip)
                lastYear = years(y);
                lastYearEndInd = y-1;
                
                tempYear = temp(lastYearStartInd:lastYearEndInd);
                precipYear = precip(lastYearStartInd:lastYearEndInd);
                
                % take the seasonal mean temperature
                tempMean(end+1) = nanmean(tempYear);
                
                % and sum of seasonal precip
                precipSum(end+1) = nansum(precipYear);
                
                % if there were any temps that exceeded the threshold,
                % stick this into the temp category
                if length(find(tempYear > tempThresh)) > 0
                    tempGroup(end+1) = years(y);

                % otherwise, stick it into the precip category
                elseif length(find(precipYear > precipThresh)) > 0
                    precipGroup(end+1) = years(y);
                end
                
                % mark the starting point of the next year
                yearInd(size(yearInd, 1)+1, 1) = y;
                yearInd(size(yearInd, 1), 2) = years(y);
                
                lastYearStartInd = y;
            end
        end

        % count how many temp/precip years we have
        tcnt = tcnt + length(tempGroup);
        pcnt = pcnt + length(precipGroup);
        
        tempGroupings = {};
        precipGroupings = {};
        
        % add temp years
        for tInd = 1:length(tempGroup)
            % find starting index of this year from the temp group
            yIndStart = find(yearInd(:, 2) == tempGroup(tInd));
            yIndEnd = yIndStart + 1;
            
            if yIndEnd < size(yearInd, 1)
                curTempData = temp(yearInd(yIndStart, 1):yearInd(yIndEnd, 1));
            else
                curTempData = temp(yearInd(yIndStart, 1):end);
            end
            
            tempGroupings{end+1} = {tempGroup(tInd), curTempData};
        end
        
        % add precip years
        for pInd = 1:length(precipGroup)
            % find starting index of this year from the temp group
            yIndStart = find(yearInd(:, 2) == precipGroup(pInd));
            yIndEnd = yIndStart + 1;
            
            if yIndEnd < size(yearInd, 1)
                curPrecipData = precip(yearInd(yIndStart, 1):yearInd(yIndEnd, 1));
            else
                curPrecipData = precip(yearInd(yIndStart, 1):end);
            end
            
            precipGroupings{end+1} = {precipGroup(pInd), curPrecipData};
        end
        
        if full
            % keep the full weather dataset
            cropWxData{3}{c}{7}{2} = {tempGroupings, precipGroupings, yearInd(:, 2), tempMean, precipSum};
        else
            % just include the grouped wx data and seasonal means
            cropWxData{3}{c}{7}{1} = {tempGroupings, precipGroupings, yearInd(:, 2), tempMean, precipSum};
        end
    end
    
    ['saving grouped data...']
    if full
        save(['2017-ag-precip/ag-data/' curFileParts{1} '-grouped-full'], 'cropWxData', '-v7.3');
    else
        save(['2017-ag-precip/ag-data/' curFileParts{1} '-grouped-small'], 'cropWxData', '-v7.3');
    end
end

File: 2017-ag-precip/ag_buildSamples.py
# -*- coding: utf-8 -*-
"""
Created on Tue Oct  3 12:10:47 2017

@author: Ethan

Separates out years with >99th percentile temperature/precip events for each county
"""

import ag_utils
import pickle
import gzip
import sys
import glob
import numpy

cropBaseDir = 'e:/data/projects/ag/crop/'

# percentile cutoffs
tempPrc = 99.95
precipPrc = 99.95

tempAbs = 32
precipAbs = 25

# get all state wx
fileList = glob.glob(cropBaseDir + '*.pgz')

groupedData = {}

# loop over all states
for file in fileList:
    mergedCropWx = {}
    
    # get the state ab from the file name
    parts = file.split('.')
    parts = parts[0].split('-')
    state = parts[-1]

    groupedData[state] = {}

    # load current state
    with gzip.open(file, 'r') as f:
        mergedCropWx = pickle.load(f)
    
    for county in mergedCropWx.keys():
        countyData = mergedCropWx[county]
        
        # skip unmatched counties
        if len(countyData) == 0:
            continue
        
        groupedData[state][county] = {'year':[], 'yield':[], 'isTempGroup':[], 'isPrecipGroup':[], \
                                      'meanTemp':[], 'totalPrecip':[], 'tempMissingFraction':[], \
                                      'precipMissingFraction':[], 'isAbsTempGroup':[], 'isAbsPrecipGroup':[]}
        
        print('processing', state + '/' + county + '...')
        
        # get indices of months in growing season
        monthInd = numpy.where((countyData['month'] >= 4) & (countyData['month'] <= 9))[0]
        
        # find indices of missing data
        notNanIndTemp = numpy.where(~numpy.isnan(countyData['temp']))[0]
        notNanIndPrecip = numpy.where(~numpy.isnan(countyData['precip']))[0]
        
        # get percentile thresholds
        tempThresh = numpy.nanpercentile(countyData['temp'], tempPrc)
        precipThresh = numpy.nanpercentile(countyData['precip'], precipPrc)
        
        # get the available years for yield, temp, and precip
        agYears = list(countyData['yield'].keys())
        wxYears = list(numpy.unique(countyData['year']))
        
        # get years available in both crop and wx data
        availYears = set(agYears).intersection(wxYears)
        
        for year in availYears:
            # find wx indicies for current year
            wxInd = numpy.where(countyData['year'] == year)[0]
            
            # find wx for growing season
            wxGrowingInd = set(monthInd.tolist()).intersection(wxInd.tolist())
            
            # and find intersection with non-nan data for temp and precip within growing season
            tempInd = list(set(notNanIndTemp.tolist()).intersection(wxGrowingInd))
            precipInd = list(set(notNanIndPrecip.tolist()).intersection(wxGrowingInd))
            
            # calculate seasonal mean temp and seasonal total precip
            seasonalMeanTemp = numpy.nan
            seasonalPrecip = numpy.nan
            # only calculate if we have indices
            if len(tempInd) > 0:
                seasonalMeanTemp = numpy.nanmean(countyData['temp'][tempInd])
            if len(precipInd) > 0:
                seasonalPrecip = numpy.nansum(countyData['precip'][precipInd])
            
            # add to grouping
            if len(tempInd) > 0 and len(precipInd) > 0:
                groupedData[state][county]['year'].append(year)
                groupedData[state][county]['yield'].append(countyData['yield'][year])
                groupedData[state][county]['meanTemp'].append(seasonalMeanTemp)
                groupedData[state][county]['tempMissingFraction'].append(1-(len(tempInd) / float(len(list(wxGrowingInd)))))
                groupedData[state][county]['totalPrecip'].append(seasonalPrecip)
                groupedData[state][county]['precipMissingFraction'].append(1-(len(precipInd) / float(len(list(wxGrowingInd)))))
                
                # if we have a temp that passes threshold
                if numpy.size(numpy.where(countyData['temp'][tempInd] >= tempThresh)) > 0:
                    groupedData[state][county]['isTempGroup'].append(True)
                else:
                    groupedData[state][county]['isTempGroup'].append(False)
                    
                if numpy.size(numpy.where(countyData['temp'][tempInd] >= tempAbs)) > 0:
                    groupedData[state][county]['isAbsTempGroup'].append(True)
                else:
                    groupedData[state][county]['isAbsTempGroup'].append(False)
                    
                # and same for precip
                if numpy.size(numpy.where(countyData['precip'][precipInd] >= precipAbs)) > 0:
                    groupedData[state][county]['isAbsPrecipGroup'].append(True)
                else:
                    groupedData[state][county]['isAbsPrecipGroup'].append(False)
        
f = open('grouped-data-995.dat', 'wb')
pickle.dump(groupedData, f)
f.close()
    
        